docs/workflow-lifecycle.md

```diff
@@ -0,0 +1,239 @@
+Frontend->>Matching: PollActivityTask
+History->>Matching: AddActivityTask (step 3 above)
+Matching->>History: RecordActivityStarted
+History->>Persistence: UpdateWorkflowExecution
+note over Persistence: append History Event: ActivityTaskStarted<br>update Mutable State, add Timer Task (activity task timeout)
+Persistence->>History: Update succeed
+History->>Matching: Record Succeed
+Matching->>Frontend: ActivityTask
+Frontend->>Worker: ActivityTask
+Worker->>Worker: Execute activity
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`RecordActivityTaskStarted` handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/handler.go#L287)
+
+</details>
+<br>
+
+---
+
+<br>
+
+5. **The Worker sends `RespondActivityTaskCompleted`, containing the activity result**
+   - A Workflow Task is added in the Matching service
+
+```mermaid
+sequenceDiagram
+Worker->>Frontend: RespondActivityTaskCompleted
+Frontend->>History: RespondActivityTaskCompleted
+History->>Persistence: UpdateWorkflowExecution
+note over Persistence: append History Events:<br>ActivityTaskCompleted  (w/ activity result), WorkflowTaskScheduled<br>update Mutable State & add Transfer Task
+Persistence->>History: Update Succeed
+History->>Frontend: Respond Succeed
+Frontend->>Worker: Respond Succeed
+loop QueueProcessor
+    History->>Persistence: GetHistoryTasks
+		History->>History: ProcessTask
+		History->>Matching: AddWorkflowTask
+end
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`RespondActivityTaskCompleted` handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/handler.go#L361)
+
+</details>
+<br>
+
+---
+
+<br>
+
+6. **The Worker dequeues the Workflow Task**
+   - It advances the workflow, and finds that it has reached its end.
+
+\<Same sequence diagram as step 2 above\>
+
+<br>
+
+---
+
+<br>
+
+7. **The Worker sends `RespondWorkflowTaskCompleted`**
+
+```mermaid
+sequenceDiagram
+Worker->>Frontend: RespondWorkflowTaskCompleted
+Frontend->>History: RespondWorkflowTaskCompleted
+History->>Persistence: UpdateWorkflowExecution
+note over Persistence: append HistoryEvents: WorkflowTaskCompleted, WorkflowExecutionCompleted<br>update Mutable State & add tasks (visibility, tiered storage, retention etc)
+Persistence->>History: Update Succeed
+History->>Frontend: Respond Succeed
+Frontend->>Worker: Respond Succeed
+loop QueueProcessor
+    History->>Persistence: GetHistoryTasks
+		History->>History: ProcessTask (Update visibility, Upload to S3, Delete data etc)
+end
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- History service [`RespondWorkflowTaskCompleted` handler](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/handler.go#L478)
+
+</details>
+<br>
+
+---
+
+<br>
+
+**Alternatively, the Activity may fail and be retried:**
```

Comment:
Should we also talk about workflow replay in this doc?

Reply:
I think that would be good. @Sushisource has some sequence diagrams which it would be good to integrate into this page. Their center of gravity is slightly more SDK, but they'd make sense together. (My aim here is to provide a foundation with basics that makes it easy for others to really improve the docs.)